File: algorithms/ar-qsrt/c/quick_sort.c
// Importing the library
#include <stdio.h>

// Function for giving the partition point
int partition(int arr[], int low, int high){  //It takes array ,high and low point and returns the pivot point
    int i = low - 1, pivot = arr[high], temp;
    // The pivot element is placed at its correct position in sorted array
    // And places all smaller (smaller than pivot) to left of pivot
    // And all greater elements to right of pivot
    for(int j = low; j < high; j++){

        if(arr[j] < pivot){
            i ++;
            temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }
    }
    temp = arr[i + 1];
    arr[i + 1] = arr[high];
    arr[high] = temp;
    return ( i+1 );

}


// Main function of quick sort algorithm (It is recursive in nature)
int quicksort(int arr[], int low, int high){    // It takes the array and high and low point
    // Initially low point is first index of array and high the last index

    if(low < high){     //Check if low is less then high
        int pivot;
         // It takes array ,high and low point and returns the pivot point
        pivot = partition(arr, low, high);
        quicksort(arr, low, pivot-1);    // Again calling the quick sort function with high point be pivot - 1
        quicksort(arr, pivot+1, high);   // Again calling the quick sort function with low point be pivot + 1
    }   // This recursion runs while low is less then high.
        // After that we get the sorted array.
}


// The main program
int main()
{
    int n;

    printf("Enter array length : ");
    scanf("%d", &n);            // Taking the length/size of the array

    int arr[n];
    printf("Enter array elements : ");
    for(int i=0;i<n;i++)
    {
        scanf("%d", &arr[i]);     // Taking array elements
    }

    quicksort(arr, 0, n-1);      // Calling the quicksort function for sorting the array in increasing order

    printf("Sorted Array is : ");
    for(int i=0;i<n;i++)
    {
        printf("%d ", arr[i]);    // Printing the sorted array
    }
    return 0;
}

File: algorithms/ar-qsrt/java/QuickSort.java
public class QuickSort {

    public static void sort(int[] arr) {
        if (arr == null || arr.length == 0) return;
        quicksort(0, arr.length - 1, arr);
    }

    private static void quicksort(int left, int right, int[] arr) {
        if (left >= right) return;

        int mid = (left + right) / 2;
        int pivot = arr[mid];

        int partitionIndex = partition(left, right, pivot, arr);
        quicksort(left, partitionIndex - 1, arr);
        quicksort(partitionIndex, right, arr);
    }

    private static int partition(int left, int right, int pivot, int[] arr) {
        int leftIndex = left;
        int rightIndex = right;

        while (leftIndex <= rightIndex) {
            while (arr[leftIndex] < pivot) {
                leftIndex++;
            }
            while (arr[rightIndex] > pivot) {
                rightIndex--;
            }
            if (leftIndex <= rightIndex) {
                int temp = arr[leftIndex];
                arr[leftIndex] = arr[rightIndex];
                arr[rightIndex] = temp;
                leftIndex++;
                rightIndex--;
            }
        }
        return leftIndex;
    }

    public static void main(String[] args) {
        int[] input = {15, 2, 8, 9, 0, 1, 12, 22};

        System.out.print("Pre-sort: ");
        for (int i: input) {
            System.out.format("%d ",i);
        }
        
        QuickSort.sort(input);

        System.out.print("Post-sort: ");
        for (int i: input) {
            System.out.format("%d ",i);
        }
    }
    
}
